hnn_core/viz.py

```diff
@@ -168,19 +168,34 @@ def plot_spikes_raster(cell_response, ax=None, show=True):
     import matplotlib.pyplot as plt
     spike_times = np.array(sum(cell_response._spike_times, []))
     spike_types = np.array(sum(cell_response._spike_types, []))
+    spike_gids = np.array(sum(cell_response._spike_gids, []))
     cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
-    spike_times_cell = [spike_times[spike_types == cell_type]
-                        for cell_type in cell_types]
+    cell_type_colors = {'L5_pyramidal': 'r', 'L5_basket': 'b',
+                        'L2_pyramidal': 'g', 'L2_basket': 'w'}
 
     if ax is None:
         fig, ax = plt.subplots(1, 1)
 
-    ax.eventplot(spike_times_cell, colors=['r', 'b', 'g', 'w'])
-    ax.legend(cell_types, ncol=2)
+    count = 0
+    for cell_type in cell_types:
+        cell_type_gids = np.unique(spike_gids[spike_types == cell_type])
+        cell_type_times, cell_type_ypos = [], []
+        for gid in cell_type_gids:
+            gid_time = spike_times[spike_gids == gid]
+            cell_type_times.append(gid_time)
+            cell_type_ypos.append(np.repeat(count, len(gid_time)))
+            count = count - 1
+
+        cell_type_times = np.concatenate(cell_type_times)
```

Comment:
```suggestion
        cell_type_times = np.array(cell_type_times)
```
? since there is only one array

Reply:
It's actually a list of arrays for each gid's spike times, just calling `np.array()` surprisingly creates an array of arrays.

I'm sure the numpy guys have a reason for that behavior... maybe

Reply:
indeed. I suppose you needed to accumulate these arrays to avoid having labels for each point in the legend? Because you could have directly called `ax.scatter` in the for loop

Reply:
Yeah that's exactly right, there is a way to manually define legend elements but I figured some awkward code is the lesser of the two evils.

Also I've found in the past that matplotlib figures are quite slow to generate when there are a lot of individual elements. I think this is because every `plt.something()` call generates a unique handle for the figure. From personal experience >100 calls to `plt.scatter()` can make take several seconds.